scipy/stats/_stats_py.py

```diff
@@ -10879,7 +10879,7 @@ def _rankdata(x, method, return_ties=False):
     elif method == 'average':
         ranks = ordinal_ranks[i] + (counts - 1)/2
     elif method == 'dense':
-        ranks = np.cumsum(i, axis=-1)[i]
+        ranks = np.cumsum(i, axis=-1, dtype=np.float64)[i]
```

Comment:
It is probably faster to work with `int` dtype where possible, and ensure conversion to `float` at the end of `_rankdata`. 

For instance:
```python3
import numpy as np
x = np.arange(10)  # same trend for larger arrays
%timeit np.cumsum(x)  # 1.58 µs ± 6.42 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
%timeit np.cumsum(x).astype(np.float64)  # 2.28 µs ± 146 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
%timeit np.cumsum(x, dtype=np.float64)  # 2.91 µs ± 20.6 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

LMK if you'd prefer that.